### PR TITLE
1356 incremental quicksaving

### LINKED ISSUE #1356: Make quicksave and autosave use rotating slots

### PR DESCRIPTION
**Pull request by FuJa0815**
I now save a binary file "last_quicksave.txt" containing the last quicksave id.

closes #1356 

Capture

### CODE REVIEW

**Comment by FuJa0815** [comment]
> 
> 
> I don't think this is a perfect solution as the player can delete that file and all other kinds of fun stuff can happen.
> So the rotating quicksaves and autosaves should instead detect the existing files based on the existing files in the saves folder based on the file names.

I thought this solution would be more efficient because it doesn't have to iterate through, maybe hundreds of files every time.

Instead of using 0 if the file does not exist, I could traverse all files to detect the last savegame

**Comment by hhyyrylainen** [comment]
My plan for saving is to anyway to move stuff to a background thread (I'm working on this currently), so the existing save lookup can run in the background while the screenshot is being taken and a frame rendered with a text saying "saving..." so, I think that we shouldn't have that external cache of the latest save file unless it takes more than like a 100 milliseconds to scan the existing save files (this would be with around 200 files as if someone has more saves it's their problem that the time taken to save is slightly increased).

File: src/saving/SaveHelper.cs [hunk 1]
```diff
@@ -131,11 +129,48 @@ private static Save CreateSaveObject(MainGameState gameState, SaveInformation.Sa
...
+            file.Open(filePath, File.ModeFlags.Read);
+            quicksaveId = file.Get32();
+            file.Close();
+        }
+        else
+        {
+            quicksaveId = 0;
+        }
+
+        return quicksaveId;*/
```
Comment (hhyyrylainen) [inline src/saving/SaveHelper.cs:167]: Commented out code should not be present.

File: src/saving/SaveHelper.cs [hunk 2]
```diff
@@ -131,11 +129,48 @@ private static Save CreateSaveObject(MainGameState gameState, SaveInformation.Sa
...
+    /// </summary>
+    private static uint GetLastQuicksaveId()
+    {
+        using (var directory = new Directory())
+        {
+            directory.Open(Constants.SAVE_FOLDER);
+            for (uint i = 0; ; i++)
+            {
+                if (!directory.FileExists(BuildQuicksaveFilename(i)))
+                    return i - 1;
```
Comment (hhyyrylainen) [inline src/saving/SaveHelper.cs:146]: I don't think this is good. Instead the existing files should be enumerated.

File: src/saving/SaveHelper.cs [hunk 3]
```diff
@@ -131,11 +129,48 @@ private static Save CreateSaveObject(MainGameState gameState, SaveInformation.Sa
         };
     }
 
+    private static string BuildQuicksaveFilename(uint quicksaveId)
+        => $"quick_save.{quicksaveId}.{Constants.SAVE_EXTENSION}";
```
Comment (hhyyrylainen) [inline src/saving/SaveHelper.cs:133]: Our style guide doesn't say anything about this, but I'm kinda leaning towards requiring methods to use curly braces and not the lambda syntax.

File: src/saving/SaveHelper.cs [hunk 4]
```diff
@@ -131,11 +129,48 @@ private static Save CreateSaveObject(MainGameState gameState, SaveInformation.Sa
...
+
+        return quicksaveId;*/
+    }
+
     private static void PerformSave(Save save, SaveInformation.SaveType type, Stopwatch stopwatch)
     {
         // TODO: implement type naming
-        var name = "quick_save." + Constants.SAVE_EXTENSION;
-
+        var name = BuildQuicksaveFilename(GetLastQuicksaveId() + 1);
```
Comment (hhyyrylainen) [inline src/saving/SaveHelper.cs:173]: The quick save numbers should be reused so that after quick save `quick_save_5` the first quick save should be overwritten.

File: src/saving/SaveHelper.cs (hunk 2 above)
Comment (hhyyrylainen) [reply src/saving/SaveHelper.cs:146]: Also the file update timestamps should be used to detect which save was last modified. Quick load should also detect autosaves and manual saves as possibly being the last one. So quick load should load the save file that has the latest timestamp despite whatever the file name might be.

File: src/saving/SaveHelper.cs [hunk 5]
```diff
@@ -131,11 +131,89 @@ private static Save CreateSaveObject(MainGameState gameState, SaveInformation.Sa
         };
     }
 
+    private static string BuildQuicksaveFilename(uint quicksaveId)
+    {
+        return $"quick_save_{quicksaveId}.{Constants.SAVE_EXTENSION}";
+    }
+
+    /// <summary>
+    ///   Returns the next quicksave-id to be used
```
Comment (hhyyrylainen) [inline src/saving/SaveHelper.cs:140]: `quicksave-id` -> `quick save id`

File: src/saving/SaveHelper.cs [hunk 6]
```diff
@@ -131,11 +131,89 @@ private static Save CreateSaveObject(MainGameState gameState, SaveInformation.Sa
         };
     }
 
+    private static string BuildQuicksaveFilename(uint quicksaveId)
```
Comment (hhyyrylainen) [inline src/saving/SaveHelper.cs:134]: I've selected in other code that `quicksave` is not one word, so you need to edit a bunch of places to be `QuickSave` as it is two separate words.

Comment (hhyyrylainen) [reply src/saving/SaveHelper.cs:134]: Also `uint` shouldn't be used here, it seems that in C# uint isn't really used, so change that to int.

File: src/saving/SaveHelper.cs [hunk 7]
```diff
@@ -131,11 +131,89 @@ private static Save CreateSaveObject(MainGameState gameState, SaveInformation.Sa
...
+        return $"quick_save_{quicksaveId}.{Constants.SAVE_EXTENSION}";
+    }
+
+    /// <summary>
+    ///   Returns the next quicksave-id to be used
+    /// </summary>
+    private static uint GetNextQuicksaveId()
+    {
+        var files = GetSaves();
+        for (uint i = 0; i < Constants.SAVE_MAX_QUICKSAVES_BEFORE_OVERRIDING_OLD_ONES; i++)
```
Comment (hhyyrylainen) [inline src/saving/SaveHelper.cs:145]: uint is also used here.

File: src/saving/SaveHelper.cs [hunk 8]
```diff
@@ -131,11 +131,89 @@ private static Save CreateSaveObject(MainGameState gameState, SaveInformation.Sa
...
+            }
+        }
+
+        // All 5 quicksaves have been used
+        var filename = GetLastSavegame(true);
+        return (uint.Parse(
+                        filename.Substr("quick_save_".Length,
+                        filename.Length - Constants.SAVE_EXTENSION.Length - "quick_save_".Length - 1),
+                        System.Globalization.CultureInfo.InvariantCulture)
+                    + 1) % (Constants.SAVE_MAX_QUICKSAVES_BEFORE_OVERRIDING_OLD_ONES + 1);
```
Comment (hhyyrylainen) [inline src/saving/SaveHelper.cs:160]: This piece of logic is way too dense and should be split over multiple statements.
Also I'm unsure if it is the best idea to parse strings back to uint. This function could perhaps return a string to make things much cleaner.

File: src/saving/SaveHelper.cs [hunk 9]
```diff
@@ -131,11 +131,89 @@ private static Save CreateSaveObject(MainGameState gameState, SaveInformation.Sa
...
+    }
+
+    private static string GetLastModifiedFile(List<string> filesToCheck)
```
Comment (hhyyrylainen) [inline src/saving/SaveHelper.cs:163]: This function should be moved to `FileHelpers` class and documented.

File: src/saving/SaveHelper.cs [hunk 10]
```diff
@@ -131,11 +131,89 @@ private static Save CreateSaveObject(MainGameState gameState, SaveInformation.Sa
...
+    {
+        return filesToCheck.ToDictionary(p => p, p =>
+        {
+            using (var file = new File())
+            {
+                return file.GetModifiedTime(PathUtils.Join(Constants.SAVE_FOLDER, p));
+            }
+        }).Aggregate((a, b) => a.Value > b.Value ? a : b).Key;
```
[the remark this answers is not in the document]
Comment (FuJa0815) [reply src/saving/SaveHelper.cs:171]: But aggregate is faster than sorting. And it makes more sense to have the filename as the key and the modified date as the value, not the other way around.

[the remark this answers is not in the document]
Comment (hhyyrylainen) [reply src/saving/SaveHelper.cs:171]: I think that code like this would work (I didn't test or write with C# support enabled so this might not compile):
```
using (var file = new File())
{
    return filesToCheck.SortBy(entry => file.GetModifiedTime(PathUtils.Join(Constants.SAVE_FOLDER, p))).First();
}
```

Which, in my opinion is much nicer than making a temporary dictionary and recreating the file object for each file.

File: Thrive.sln [hunk 11]
```diff
@@ -1,5 +1,7 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 2012
+# Visual Studio Version 16
```
Comment (hhyyrylainen) [inline Thrive.sln:2]: Still changes in this file

File: assets/models/organelles/Flagellum.glb.import [hunk 12]
```diff
@@ -24,6 +24,7 @@ meshes/ensure_tangents=true
 meshes/storage=0
 meshes/light_baking=0
 meshes/lightmap_texel_size=0.1
+skins/use_named_skins=true
```
Comment (hhyyrylainen) [inline assets/models/organelles/Flagellum.glb.import:27]: Have you tested this change extensively? This is related to: https://github.com/Revolutionary-Games/Thrive/issues/1259

File: project.godot [hunk 13]
```diff
@@ -205,9 +205,15 @@ screenshot={
 
 file_logging/enable_file_logging=true
 
+[mono]
+
+debugger_agent/wait_for_debugger=true
```
Comment (hhyyrylainen) [inline project.godot:210]: Also changes here

File: src/saving/FileHelpers.cs [hunk 14]
```diff
@@ -18,7 +20,7 @@ public static void MakeSureDirectoryExists(string path)
 
             if (result != Error.AlreadyExists && result != Error.Ok)
             {
-                throw new IOException($"can't create folder: {path}");
+                throw new System.IO.IOException($"can't create folder: {path}");
```
Comment (hhyyrylainen) [inline src/saving/FileHelpers.cs:23]: Why did you remove the System.IO import?

File: src/saving/FileHelpers.cs [hunk 15]
```diff
@@ -36,4 +38,24 @@ public static bool DeleteFile(string path)
...
     }
+
+    /// <param name="filesToCheck">
+    ///   The files to check. Full path needed.
+    /// </param>
+    /// <returns>
+    ///   Returns the last modified file in the list.
+    /// </returns>
+    /// <summary>
+    ///   Gets the last modified file.
```
Comment (hhyyrylainen) [inline src/saving/FileHelpers.cs:49]: Summary should be first.

File: src/saving/FileHelpers.cs [hunk 16]
```diff
@@ -36,4 +38,24 @@ public static bool DeleteFile(string path)
...
+    /// </summary>
+    public static string GetLastModifiedFile(IEnumerable<string> filesToCheck)
+    {
+        return filesToCheck.OrderBy(GetModifiedDate).FirstOrDefault();
```
Comment (hhyyrylainen) [inline src/saving/FileHelpers.cs:53]: There is still the problem of making a new File object for *each* iteration, and disposing it also in each iteration.

File: src/saving/SaveHelper.cs [hunk 17]
```diff
@@ -131,11 +131,96 @@ private static Save CreateSaveObject(MainGameState gameState, SaveInformation.Sa
...
+        return $"quick_save_{quickSaveId}.{Constants.SAVE_EXTENSION}";
+    }
+
+    /// <summary>
+    ///   Returns the next quick save id to be used
+    /// </summary>
+    private static int GetNextQuickSaveId()
+    {
+        var files = GetSaves();
+        var enumerable = files as string[] ?? files.ToArray();
```
Comment (hhyyrylainen) [inline src/saving/SaveHelper.cs:145]: I'm suspicious of this line. Why not just `GetSaves().ToList()`?

File: src/saving/SaveHelper.cs [hunk 18]
```diff
@@ -131,11 +131,96 @@ private static Save CreateSaveObject(MainGameState gameState, SaveInformation.Sa
...
+        var filename = GetLastSaveGame(true);
+
+        // Get the id out of the quick save file name.
+        filename = filename.Substr("quick_save_".Length,
+            filename.Length - Constants.SAVE_EXTENSION.Length - "quick_save_".Length - 1);
+
+        var id = int.Parse(filename, System.Globalization.CultureInfo.InvariantCulture);
+
+        // Get the next id. Overflow if MaxQuickSavesBeforeOverriding is reached.
+        return (id + 1) % (Settings.Instance.MaxQuickSavesBeforeOverriding + 1);
```
Comment (hhyyrylainen) [inline src/saving/SaveHelper.cs:165]: I still don't like this part. What I'd prefer is if you did something like:
- get full paths from files with a map, and get oldest modified file, and return that filename

File: src/saving/SaveHelper.cs [hunk 19]
```diff
@@ -131,11 +131,96 @@ private static Save CreateSaveObject(MainGameState gameState, SaveInformation.Sa
...
+            saves = saves.Where(p => p.StartsWith("quick_save_", StringComparison.Ordinal));
+        }
+
+        // Relative path to absolute path
+        saves = saves.Select(p => PathUtils.Join(Constants.SAVE_FOLDER, p));
+
+        var fullPath = FileHelpers.GetLastModifiedFile(saves);
+
+        // Get filename of fullPath
+        return fullPath.Substring(fullPath.LastIndexOf(PathUtils.PATH_SEPARATOR) + 1);
```
Comment (hhyyrylainen) [inline src/saving/SaveHelper.cs:189]: This should be made into an utility function in `PathUtils`

File: src/saving/FileHelpers.cs (hunk 14 above)
Comment (FuJa0815) [reply src/saving/FileHelpers.cs:23]: Because System.IO has a File class.
I think, that the Godot.File class will be used more frequently than System.IO.File in this class

Comment (hhyyrylainen) [reply src/saving/FileHelpers.cs:23]: Okay, that's fine.

File: src/saving/FileHelpers.cs (hunk 16 above)
Comment (FuJa0815) [reply src/saving/FileHelpers.cs:53]: but I need to make it somewhere to get the date

Comment (hhyyrylainen) [reply src/saving/FileHelpers.cs:53]: Create the File object before, and use it in the loop.
